File: llm_engine/src/db_schema.py
import os
import logging
import psycopg2
from typing import Dict, Any, List
from dotenv import load_dotenv

load_dotenv()

logger = logging.getLogger(__name__)

async def get_database_schema() -> Dict[str, Any]:
    """
    Retrieve the database schema from the PostgreSQL database
    Returns a dictionary representation of tables and columns
    """
    db_url = os.getenv("DATABASE_URL")
    if not db_url:
        logger.warning("DATABASE_URL not set, using hardcoded schema")
        return get_hardcoded_schema()
    
    try:
        conn = psycopg2.connect(db_url)
        cursor = conn.cursor()
        
        # Query to get all tables
        cursor.execute("""
            SELECT tablename FROM pg_catalog.pg_tables
            WHERE schemaname = 'public';
        """)
        
        tables = cursor.fetchall()
        schema = {}
        
        for (table_name,) in tables:
            # Query to get columns for this table
            cursor.execute(f"""
                SELECT column_name, data_type, is_nullable
                FROM information_schema.columns
                WHERE table_name = '{table_name}';
            """)
            
            columns = []
            for column_name, data_type, is_nullable in cursor.fetchall():
                columns.append({
                    "name": column_name,
                    "type": data_type,
                    "nullable": is_nullable == "YES"
                })
            
            schema[table_name] = {
                "columns": columns
            }
        
        conn.close()
        return schema
        
    except Exception as e:
        logger.error(f"Error fetching database schema: {str(e)}")
        # Fall back to hardcoded schema
        return get_hardcoded_schema()

def get_hardcoded_schema() -> Dict[str, Any]:
    """
    Return a hardcoded schema for the cars table
    Used as fallback when database connection fails
    """
    return {
        "cars": {
            "columns": [
                {"name": "id", "type": "integer", "nullable": False},
                {"name": "model", "type": "varchar(50)", "nullable": False},
                {"name": "mpg", "type": "numeric(5,1)", "nullable": True},
                {"name": "cyl", "type": "integer", "nullable": True},
                {"name": "disp", "type": "numeric(6,1)", "nullable": True},
                {"name": "hp", "type": "integer", "nullable": True},
                {"name": "drat", "type": "numeric(4,2)", "nullable": True},
                {"name": "wt", "type": "numeric(5,3)", "nullable": True},
                {"name": "qsec", "type": "numeric(5,2)", "nullable": True},
                {"name": "vs", "type": "integer", "nullable": True},
                {"name": "am", "type": "integer", "nullable": True},
                {"name": "gear", "type": "integer", "nullable": True},
                {"name": "carb", "type": "integer", "nullable": True},
            ]
        }
    }

File: llm_engine/src/db_schema.rs
use std::env;
use anyhow::Result;
use tracing::{info, error};
use serde_json::{Value, json};
use std::collections::HashMap;
use tokio_postgres::NoTls;

/// Retrieve the database schema from the PostgreSQL database
/// Returns a dictionary representation of tables and columns
pub async fn get_database_schema() -> Result<HashMap<String, Value>> {
    let db_url = match env::var("DATABASE_URL") {
        Ok(url) => url,
        Err(_) => {
            info!("DATABASE_URL not set, using hardcoded schema");
            return Ok(get_hardcoded_schema());
        }
    };
    
    match get_schema_from_db(&db_url).await {
        Ok(schema) => Ok(schema),
        Err(e) => {
            error!("Error fetching database schema: {}", e);
            info!("Falling back to hardcoded schema");
            Ok(get_hardcoded_schema())
        }
    }
}

/// Query the database to get its schema
async fn get_schema_from_db(db_url: &str) -> Result<HashMap<String, Value>> {
    // Connect to the database
    let (client, connection) = tokio_postgres::connect(db_url, NoTls).await?;
    
    // The connection object performs the actual communication with the database,
    // so spawn it off to run on its own
    tokio::spawn(async move {
        if let Err(e) = connection.await {
            error!("connection error: {}", e);
        }
    });
    
    // Query to get all tables
    let tables_query = "
        SELECT tablename FROM pg_catalog.pg_tables
        WHERE schemaname = 'public';
    ";
    
    let mut schema = HashMap::new();
    
    for row in client.query(tables_query, &[]).await? {
        let table_name: String = row.get(0);
        
        // Query to get columns for this table
        let columns_query = format!(
            "
                SELECT column_name, data_type, is_nullable
                FROM information_schema.columns
                WHERE table_name = '{}';
            ",
            table_name
        );
        
        let mut columns = Vec::new();
        
        for col_row in client.query(&columns_query, &[]).await? {
            let column_name: String = col_row.get(0);
            let data_type: String = col_row.get(1);
            let is_nullable: String = col_row.get(2);
            
            columns.push(json!({
                "name": column_name,
                "type": data_type,
                "nullable": is_nullable == "YES"
            }));
        }
        
        schema.insert(table_name, json!({
            "columns": columns
        }));
    }
    
    Ok(schema)
}

/// Return a hardcoded schema for the cars table
/// Used as fallback when database connection fails
fn get_hardcoded_schema() -> HashMap<String, Value> {
    let mut schema = HashMap::new();
    
    let columns = json!([
        {"name": "id", "type": "integer", "nullable": false},
        {"name": "model", "type": "varchar(50)", "nullable": false},
        {"name": "mpg", "type": "numeric(5,1)", "nullable": true},
        {"name": "cyl", "type": "integer", "nullable": true},
        {"name": "disp", "type": "numeric(6,1)", "nullable": true},
        {"name": "hp", "type": "integer", "nullable": true},
        {"name": "drat", "type": "numeric(4,2)", "nullable": true},
        {"name": "wt", "type": "numeric(5,3)", "nullable": true},
        {"name": "qsec", "type": "numeric(5,2)", "nullable": true},
        {"name": "vs", "type": "integer", "nullable": true},
        {"name": "am", "type": "integer", "nullable": true},
        {"name": "gear", "type": "integer", "nullable": true},
        {"name": "carb", "type": "integer", "nullable": true}
    ]);
    
    schema.insert("cars".to_string(), json!({
        "columns": columns
    }));
    
    schema
}
